From b4dfe00c5ad4aa0acf7f55e8a84e3f1590425c40 Mon Sep 17 00:00:00 2001
From: Andy Shevchenko <andriy.shevchenko@linux.intel.com>
Date: Fri, 15 Jun 2018 14:06:00 +0300
Subject: [PATCH 1562/2313] FIXUP: mfd: intel-lpss: Probe UART devices
 synchronously

From Sathyanarayanan Kuppuswamy:

Issue: With this patch merged, I am not able to get my console working
sometimes ( ~70% of the time). Without this patch console has no issues.

Signed-off-by: Andy Shevchenko <andriy.shevchenko@linux.intel.com>
---
 drivers/mfd/intel-lpss.c | 20 +++++++++++++++++---
 1 file changed, 17 insertions(+), 3 deletions(-)

diff --git a/drivers/mfd/intel-lpss.c b/drivers/mfd/intel-lpss.c
index d25d6ae..d6aa792 100644
--- a/drivers/mfd/intel-lpss.c
+++ b/drivers/mfd/intel-lpss.c
@@ -385,9 +385,8 @@ static void intel_lpss_unregister_clock(struct intel_lpss *lpss)
 	intel_lpss_unregister_clock_tree(lpss->clk);
 }
 
-static void intel_lpss_async_add_devices(void *_lpss, async_cookie_t cookie)
+static int intel_lpss_add_devices(struct intel_lpss *lpss)
 {
-	struct intel_lpss *lpss = _lpss;
 	int ret;
 
 	if (intel_lpss_has_idma(lpss)) {
@@ -416,6 +415,12 @@ static void intel_lpss_async_add_devices(void *_lpss, async_cookie_t cookie)
 		intel_lpss_ltr_hide(lpss);
 		intel_lpss_unregister_clock(lpss);
 	}
+	return ret;
+}
+
+static void intel_lpss_async_add_devices(void *lpss, async_cookie_t cookie)
+{
+	intel_lpss_add_devices(lpss);
 }
 
 int intel_lpss_probe(struct device *dev,
@@ -464,7 +469,16 @@ int intel_lpss_probe(struct device *dev,
 	if (ret)
 		dev_warn(dev, "Failed to create debugfs entries\n");
 
-	async_schedule(intel_lpss_async_add_devices, lpss);
+	/*
+	 * Probe UART devices synchronously to avoid serial interface
+	 * enumeration unpredictability.
+	 */
+	if (lpss->type == LPSS_DEV_UART) {
+		ret = intel_lpss_add_devices(lpss);
+		if (ret)
+			goto err_clk_register;
+	} else
+		async_schedule(intel_lpss_async_add_devices, lpss);
 
 	return 0;
 
-- 
2.7.4

